packages/eas-cli/src/project/publish.ts

```diff
@@ -0,0 +1,308 @@
+}
+
+export function getBase64URLEncoding(buffer: Buffer): string {
+  const base64 = buffer.toString('base64');
+  return base64.replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
+}
+
+/**
+ * The storage key is used to store the asset in GCS
+ */
+export function getStorageKey(contentType: string, contentHash: string): string {
+  const nullSeparator = Buffer.alloc(1);
+  const hash = crypto
+    .createHash('sha256')
+    .update(contentType)
+    .update(nullSeparator)
+    .update(contentHash)
+    .digest();
+  return getBase64URLEncoding(hash);
+}
+
+/**
+ * Convenience function that computes an assets storage key starting from its buffer.
+ */
+export function getStorageKeyForAsset(asset: RawAsset): string {
+  const fileHash = getBase64URLEncoding(
+    crypto.createHash('sha256').update(asset['buffer']).digest()
+  );
+  const contentType = asset['contentType'];
+  return getStorageKey(contentType, fileHash);
+}
+
+export function convertAssetToUpdateInfoGroupFormat(asset: RawAsset): PartialManifestAsset {
+  const fileHash = getBase64URLEncoding(
+    crypto.createHash('sha256').update(asset['buffer']).digest()
+  );
+  const contentType = asset['contentType'];
+  const storageKey = getStorageKey(contentType, fileHash);
+  const bundleKey = [
+    crypto.createHash('md5').update(asset['buffer']).digest('hex'),
+    asset.type,
+  ].join('.');
+
+  return {
+    fileHash,
+    contentType,
+    storageBucket: STORAGE_BUCKET,
+    storageKey,
+    bundleKey,
+  };
+}
+
+export function buildUpdateInfoGroup(assets: CollectedAssets) {
+  let platform: PublishPlatforms;
+  const updateInfoGroup: any = {};
+  for (platform in assets) {
+    updateInfoGroup[platform] = {
+      launchAsset: convertAssetToUpdateInfoGroupFormat(assets[platform]?.launchAsset!),
+      assets: assets[platform]?.assets!.map(convertAssetToUpdateInfoGroupFormat),
+    };
+  }
+  return updateInfoGroup;
+}
+
+export function getDistRoot(customDist: string): string {
+  const projectRoot = process.cwd();
+  const distRoot = path.join(projectRoot, customDist);
+  if (!fs.existsSync(distRoot)) {
+    throw new Error(`${distRoot} does not exist. Please create it with your desired bundler.`);
+  }
+  return distRoot;
+}
+
+export function collectUserDefinedAssets(distRoot: string): RawAsset[] {
+  const assetRoot = path.join(distRoot, 'assets');
+  const assetPointers = Platforms.map(platform => {
+    const assetJsonPath = path.join(distRoot, `${platform}-index.json`);
+    // load JSON with fs instead of require for easier mocking in tests
+    return JSON.parse(fs.readFileSync(assetJsonPath).toString()).bundledAssets;
+  }).flat();
+
+  return [...new Set(assetPointers)].map(pointer => {
+    const [filename, ext] = pointer.split('_').pop().split('.');
+    return {
+      type: ext ?? '',
+      contentType: guessContentTypeFromExtension(ext),
+      buffer: fs.readFileSync(path.join(assetRoot, filename)),
```

Comment:
Great points. 
* I changed it to pass around the `filename` instead of a `buffer`.
* Uploads and hashing are parallelized.
* Not sure how to handle progress bar for multiple uploads. Left it out for now.